### PR TITLE
repository: added cleanup for PlainClone. Fixes #741

### LINKED ISSUE #741: After failing PlainClone .git directory nevertheless exists
### What version of Go are you using (`go version`)?
go version go1.9.3 darwin/amd64

### What did you do?

```go
package main

import (
	"log"
	"os"
	"gopkg.in/src-d/go-git.v4"
	"gopkg.in/src-d/go-git.v4/plumbing"
)

func main() {
	_, err := git.PlainClone("repository", false, &git.CloneOptions{
		URL:           "https://github.com/git-fixtures/basic.git",
		ReferenceName: plumbing.ReferenceName("notexistingbranch"),
	})
	if err != nil {
		if _, err := os.Stat("repository/.git"); err != nil {
			if os.IsNotExist(err) {
				log.Print("Not Exists")
			} else {
			}
		}
		log.Print("Exists")
	}
}
```

### What did you expect to see?
I expexted to see, that the .git directory does not exist after a failing pull

### What did you see instead?
I saw that the directory exist, when the error is not nil.

### PR DESCRIPTION
**Pull request by bjaroszewski**
This PR introduces the cleanup for the PlainClone().

While using PlainClone, if the clone will raise an error (f.e. incorrect credentials used for authentication) we have work dir left.

### CODE REVIEW

File: repository.go [hunk 1]
```diff
@@ -334,7 +335,23 @@ func dotGitFileToOSFilesystem(path string, fs billy.Filesystem) (bfs billy.Files
 // if the new repository will be bare or normal. If the path is not empty
 // ErrRepositoryAlreadyExists is returned.
 func PlainClone(path string, isBare bool, o *CloneOptions) (*Repository, error) {
-	return PlainCloneContext(context.Background(), path, isBare, o)
+	r, err := PlainCloneContext(context.Background(), path, isBare, o)
+	if err != nil {
```
Comment (smola) [inline repository.go]: Cleanup logic should be the same for `PlainCloneContext`

File: repository.go [hunk 2]
```diff
@@ -334,7 +335,23 @@ func dotGitFileToOSFilesystem(path string, fs billy.Filesystem) (bfs billy.Files
...
+		case transport.ErrAuthorizationFailed:
+			os.RemoveAll(path)
+			return nil, err
+		case ErrRepositoryAlreadyExists:
+			return nil, err
+		default:
+			os.Remove(path)
+			return nil, err
+		}
+	}
```
Comment (smola) [inline repository.go]: nitpick: add a space here (we usually add a space after every if/for block.

File: repository.go [hunk 3]
```diff
@@ -334,7 +335,23 @@ func dotGitFileToOSFilesystem(path string, fs billy.Filesystem) (bfs billy.Files
...
+		case transport.ErrAuthenticationRequired:
+			os.RemoveAll(path)
+			return nil, err
+		case transport.ErrAuthorizationFailed:
+			os.RemoveAll(path)
+			return nil, err
+		case ErrRepositoryAlreadyExists:
+			return nil, err
+		default:
+			os.Remove(path)
```
Comment (smola) [inline repository.go]: Why the default case is `Remove` and auth errors are `RemoveAll`?

File: repository.go [hunk 4]
```diff
@@ -334,7 +334,13 @@ func dotGitFileToOSFilesystem(path string, fs billy.Filesystem) (bfs billy.Files
 // if the new repository will be bare or normal. If the path is not empty
 // ErrRepositoryAlreadyExists is returned.
 func PlainClone(path string, isBare bool, o *CloneOptions) (*Repository, error) {
-	return PlainCloneContext(context.Background(), path, isBare, o)
+	r, err := PlainCloneContext(context.Background(), path, isBare, o)
```
Comment (smola) [inline repository.go]: Keep it in a single place. If you implement it in `PlainCloneContext`, then that's it, you don't need it here.

File: repository_test.go [hunk 5]
```diff
@@ -577,6 +577,24 @@ func (s *RepositorySuite) TestPlainCloneContext(c *C) {
...
+
+	r, err := PlainCloneContext(ctx, dir, false, &CloneOptions{
+		URL: "incorrectOnPurpose",
+	})
+	c.Assert(r, IsNil)
+	c.Assert(err, NotNil)
+
+	_, err = os.Stat(dir)
+	dirNotExist := os.IsNotExist(err)
+	c.Assert(dirNotExist, Equals, true)
```
Comment (smola) [inline repository_test.go]: The behaviour does not seem correct: it removes a directory that already existed before. go-git shouldn't remove anything it didn't create during clone.

File: repository_test.go [hunk 6]
```diff
@@ -577,6 +578,23 @@ func (s *RepositorySuite) TestPlainCloneContext(c *C) {
...
+	tmpDir := c.MkDir()
+	repoDir := path.Join(path.Dir(tmpDir), "repoDir")
+	r, err := PlainCloneContext(ctx, repoDir, false, &CloneOptions{
+		URL: "incorrectOnPurpose",
+	})
+	c.Assert(r, IsNil)
+	c.Assert(err, NotNil)
+
+	_, err = os.Stat(repoDir)
+	dirNotExist := os.IsNotExist(err)
```
Comment (smola) [inline repository_test.go]: There should be another test to check that if the directory exists, it is not removed by PlainClone.

**Comment by bjaroszewski** [comment]
@mcuadros as per discussion on the slack, I will introduce the changes as @smola suggested for this cleanup on PlainCloneContext(). I should have some free evening this week, so I'll update the PR. Sorry for the delay with it.

File: repository.go [hunk 7]
```diff
@@ -347,12 +348,64 @@ func PlainClone(path string, isBare bool, o *CloneOptions) (*Repository, error)
 // operation is complete, an error is returned. The context only affects to the
 // transport operations.
 func PlainCloneContext(ctx context.Context, path string, isBare bool, o *CloneOptions) (*Repository, error) {
+	dirEmpty := false
+	dirExist := false
+
+	file, err := os.Stat(path)
+	pathNotExist := os.IsNotExist(err)
```
Comment (smola) [inline repository.go]: You can write here:

```
file, err := os.Stat(path)
if os.IsNotExist(err) {
  dirExist = true
} else if err != nil {
  return nil, err
}
```

If `os.Stat`returns an error and it is not a not exist error, something is wrong (e.g. permissions) and we should not continue.

File: repository.go [hunk 8]
```diff
@@ -347,12 +348,64 @@ func PlainClone(path string, isBare bool, o *CloneOptions) (*Repository, error)
...
+	if dirExist {
+		fh, err := os.Open(path)
+		if err != nil {
+			return nil, err
+		}
+		defer fh.Close()
+
+		names, err := fh.Readdirnames(1)
+
+		if err == io.EOF && len(names) == 0 {
```
Comment (smola) [inline repository.go]: Here if `err` is not nil or `io.EOF`, an error should be returned.

File: repository.go [hunk 9]
```diff
@@ -347,12 +348,64 @@ func PlainClone(path string, isBare bool, o *CloneOptions) (*Repository, error)
...
+				return nil, err
+			}
+
+			for _, name := range names {
+				err = os.RemoveAll(filepath.Join(path, name))
+				if err != nil {
+					return nil, err
+				}
+			}
+		} else if !dirExist {
```
Comment (smola) [inline repository.go]: Avoid additional indentation by reversing condition order:

```
if !dirExist {
    os.RemoveAll(path)
    return nil, err
}

// continue here

File: repository_test.go [hunk 10]
```diff
@@ -579,6 +579,38 @@ func (s *RepositorySuite) TestPlainCloneContext(c *C) {
 	c.Assert(err, NotNil)
 }
 
+func (s *RepositorySuite) TestPlainCloneContextWithIncorrectRepo(c *C) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	tmpDir := c.MkDir()
+	repoDir := filepath.Join(tmpDir, "repoDir") //path.Join(path.Dir(tmpDir), "repoDir")
```
Comment (smola) [inline repository_test.go]: Remove commented out code.

**Comment by smola** [comment]
CI fails:

```
# gopkg.in/src-d/go-git.v4
./repository.go:373:16: undefined: ErrDirNotEmpty
```

File: repository.go (hunk 7 above)
Comment (smola) [reply repository.go]: Sorry, my snippet was wrong.
Leave your logic as it is if behavior with permission errors is correct.
Do not use `pathNotExist` variable here only to use it for the if condition. Just write `if !os.IsNotExist(err)`.

File: repository.go (hunk 9 above)
Comment (smola) [reply repository.go]: Meh, I confused `dirExist` and  `dirEmpty`, sorry.

File: repository_test.go [hunk 11]
```diff
@@ -576,7 +576,7 @@ func (s *RepositorySuite) TestPlainCloneContextWithProperParameters(c *C) {
 		URL: s.GetBasicLocalRepositoryURL(),
 	})
 
-	c.Assert(r, NoNill)
+	c.Assert(r, NotNill)
```
Comment (PaluMacil) [inline repository_test.go]: still a typo here... you want NotNil, not NotNill

**Pull request closed without merging**